Sources/Conduit/Core/DarwinNotificationCenter.swift

```diff
@@ -0,0 +1,130 @@
+//
+//  DarwinNotificationCenter.swift
+//  Conduit
+//
+//  Created by John Hammerlund on 6/12/18.
+//  Copyright © 2018 MINDBODY. All rights reserved.
+//
+
+import Foundation
+
+class DarwinNotificationObserver {
+
+    let notificationName: String
+    let handler: (DarwinNotificationObserver) -> Void
+    let queue: DispatchQueue
+
+    init(notificationName: String, queue: DispatchQueue, handler: @escaping (DarwinNotificationObserver) -> Void) {
+        self.notificationName = notificationName
+        self.queue = queue
+        self.handler = handler
+    }
+
+}
+
+/// Utilizes Core OS (Darwin) notifications, which are delivered system-wide. These notifications cannot carry
+/// payloads, and sensitive information should not be broadcasted through this system.
+class DarwinNotificationCenter {
+
+    /// A Darwin system notification
+    struct Notification {
+        let name: String
+
+        init(_ name: String) {
+            self.name = name
+        }
+    }
+
+    static let `default` = DarwinNotificationCenter()
+    private let notifierQueue = DispatchQueue(label: "com.mindbodyonline.Conduit.DarwinNotificationCenter", attributes: .concurrent)
+    private var observerMap: [String: [DarwinNotificationObserver]] = [:]
+
+    /// Registers a new observer for a given system notification
+    ///
+    /// - Parameters:
+    ///   - notification: The notification to observe
+    ///   - operationQueue: The queue where observations should be handled (defaults to a background queue)
+    ///   - handler: Handles observations when they occur
+    /// - Returns: A registered observer
+    @discardableResult
+    func registerObserver(notification: Notification, queue: DispatchQueue = .global(), handler: @escaping (DarwinNotificationObserver) -> Void)
+        -> DarwinNotificationObserver {
+        let notificationName = notification.name
+        let observer = DarwinNotificationObserver(notificationName: notificationName, queue: queue, handler: handler)
+        notifierQueue.async(flags: .barrier) {
+            var observers = self.observerMap[notificationName] ?? []
+            let existingKeys = self.observerMap.keys
+            observers.append(observer)
+            self.observerMap[notificationName] = observers
+            if existingKeys.contains(notificationName) {
+                return
+            }
+
+            let center = CFNotificationCenterGetDarwinNotifyCenter()
+            /// This will fire on the main thread
```

Comment:
is that built into the object?

Reply:
It's documented in the Discussion portion of the function documentation:
> The Darwin Notify Center has no notion of per-user sessions, all notifications are system-wide. As with distributed notifications, the main thread's run loop must be running in one of the common modes (usually kCFRunLoopDefaultMode) for Darwin-style notifications to be delivered.

Reply:
Interesting, thats good to know. Thanks!